content/algorithms/pagerank/pagerank.md

```diff
@@ -0,0 +1,513 @@
+)
+
+plt.title("G2", y=-0.1)
+plt.show()
+```
+
+In the above Markov chain, when we apply transitions, the distribution never converges and only oscillates. This is called a <b><i>periodic</i></b> Markov chain.
+
+Periodic Markov chains are characterized by having subsets of states that form closed loops or cycle, where transitions within the cycle are more probable than transitions between cycles. These loops introduce a regularity in the chain's behavior, causing the system to exhibit repetitive patterns as it evolves over time.
+
+A periodic Markoc chain does not converge to a stationary distribution.
+
++++
+
+Thus, we can conclude that for a Markov chain to have a unique stationary distribution, it must satisfy two conditions:
+
+<b>Reducibility (Uniqueness problem):</b> For a unique stationary distribution $\pi$ to exist, the Markov chain must not be reducible. The Markov chain must be irreducible i.e it is possible to reach any state from any other state, either in a single step or over multiple steps.
+
+<b>Periodicity (Convergence problem):</b> For all initial distributions $\pi_{0}$ to converge to $\pi$, the Markov chain must not be periodic. The Markov chain must be aperiodic i.e. when we apply transitions, the distribution across states should not oscillate but converge after sufficient time.
+
+When a Markov chain is both irreducible and aperiodic, it is guaranteed to have a unique stationary distribution. This distribution represents the long-term behavior of the system, where the probabilities of being in each state stabilize and remain constant over time.
+
++++
+
+## PageRank Algorithm
+
+The Pagerank algorithm is fundamentally based on the principle of computing Stationary Distributions of Markov chains. Our methods for computing stationary distributions rely on the assumption that the web network is a reducible and aperiodic Markov chain. However it may not always be true in reality.
+
+There maybe webpages that have no outgoing links, so all users will be absorbed into this state after some transitions (these are called sink states). There maybe two or more webpages linked to each other such that once a user enters, they get stuck in a never ending cycle and all users eventually end up cycling between these two states.
+
+<i> So, how do we handle Markov chains that don't satisfy the required conditions?</i>
+
+At every step, we transition to a random state with the probability $\frac{1-\alpha}{N}$ where $N = $ Number of states and we can pick any value for $\alpha$.
+
+But it is not sufficient to just add these random transition probabilities because it breaks the assumption of a probability distribution of across states since the distribution no longer sums to one. To solve this problem, we use the damping factor $\alpha$. We first reduce the original transition probability by $\alpha$ and then add the random state transition probability of $\frac{1-\alpha}{N}$.
```

Comment:
```suggestion
```suggestion
We change the random walk ever so slightly. At every step, take a random walk, as described above, with chance $\alpha$, and with chance $(1-\alpha)$ we "jump" to a uniformly chosen random state with the probability $\frac{1-\alpha}{N}$ for each of the $N$ states. Typically $\alpha$ is chosen to be about 0.85. The jumping ensures that any state can be reached from any other (irreducible) and that the jumps can happen at any timestep (aperiodic). Note that the sum of the probabilities is 1 as it should be.
```
```